File: ios/Classes/BleHealthDevicePlugin.swift
import Flutter
import UIKit
import SmartBLEHealthDeviceSDK

public class SwiftBleHealthDevicePlugin: NSObject, FlutterPlugin {
    private let manager: HealthBLEManager

    // Initialize with a default DeviceConfiguration
    public override init() {
        let config = [
            DeviceConfiguration(
                deviceId: UUID(),
                deviceType: .glucose,
                allowedOperations: [.read, .write, .notify]
            )
        ]
        manager = HealthBLEManager(configurations: config)
        super.init()
    }

    // Register the plugin with Flutter
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "ble_health_device_plugin", binaryMessenger: registrar.messenger())
        let instance = SwiftBleHealthDevicePlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    // Handle Method Channel calls
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)

        case "startScanning":
            let arguments = call.arguments as? [String: Any]
            let serviceUuids = (arguments?["serviceUuids"] as? [String] ?? []).compactMap { UUID(uuidString: $0) }
            manager.startScanning { scanResult in
                switch scanResult {
                case .success(let devices):
                    let deviceList = devices.map { [
                        "id": $0.id.uuidString,
                        "name": $0.name,
                        "isConnected": $0.isConnected
                    ] }
                    result(deviceList)
                case .failure(let error):
                    result(FlutterError(code: "SCAN_FAILED", message: error.localizedDescription, details: nil))
                }
            }

        case "connect":
            if let arguments = call.arguments as? [String: Any],
               let deviceIdString = arguments["deviceId"] as? String,
               let deviceId = UUID(uuidString: deviceIdString) {
                manager.connect(to: deviceId) { connectResult in
                    switch connectResult {
                    case .success:
                        result(nil)
                    case .failure(let error):
                        result(FlutterError(code: "CONNECT_FAILED", message: error.localizedDescription, details: nil))
                    }
                }
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Device ID is required or invalid", details: nil))
            }

        case "disconnect":
            if let arguments = call.arguments as? [String: Any],
               let deviceIdString = arguments["deviceId"] as? String,
               let deviceId = UUID(uuidString: deviceIdString) {
                manager.disconnect(from: deviceId) { disconnectResult in
                    switch disconnectResult {
                    case .success:
                        result(nil)
                    case .failure(let error):
                        result(FlutterError(code: "DISCONNECT_FAILED", message: error.localizedDescription, details: nil))
                    }
                }
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Device ID is required or invalid", details: nil))
            }

        case "readCharacteristic":
            if let arguments = call.arguments as? [String: Any],
               let deviceIdString = arguments["deviceId"] as? String,
               let serviceUuidString = arguments["serviceUuid"] as? String,
               let characteristicUuid = arguments["characteristicUuid"] as? String,
               let deviceId = UUID(uuidString: deviceIdString) {
                let serviceUuid = CBUUID(string: serviceUuidString)
                manager.readCharacteristic(deviceId: deviceId, serviceUUID: serviceUuid, characteristicUUID: characteristicUuid) { dataResult in
                    switch dataResult {
                    case .success(let healthData):
                        result([
                            "type": healthData.type == .heartRate ? "HEART_RATE" : "GLUCOSE",
                            "value": healthData.value,
                            "timestamp": Int64(healthData.timestamp.timeIntervalSince1970 * 1000)
                        ])
                    case .failure(let error):
                        result(FlutterError(code: "READ_FAILED", message: error.localizedDescription, details: nil))
                    }
                }
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing or invalid arguments", details: nil))
            }

        case "writeCharacteristic":
            if let arguments = call.arguments as? [String: Any],
               let deviceIdString = arguments["deviceId"] as? String,
               let serviceUuidString = arguments["serviceUuid"] as? String,
               let characteristicUuid = arguments["characteristicUuid"] as? String,
               let typeString = arguments["type"] as? String,
               let value = arguments["value"] as? Int,
               let timestamp = arguments["timestamp"] as? Int64,
               let deviceId = UUID(uuidString: deviceIdString) {
                let type = typeString == "HEART_RATE" ? HealthDataType.heartRate : HealthDataType.glucose
                let healthData = HealthData(type: type, value: value, timestamp: Date(timeIntervalSince1970: TimeInterval(timestamp) / 1000))
                let serviceUuid = CBUUID(string: serviceUuidString)
                manager.writeCharacteristic(deviceId: deviceId, serviceUUID: serviceUuid, characteristicUUID: characteristicUuid, data: healthData) { writeResult in
                    switch writeResult {
                    case .success:
                        result(nil)
                    case .failure(let error):
                        result(FlutterError(code: "WRITE_FAILED", message: error.localizedDescription, details: nil))
                    }
                }
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing or invalid arguments", details: nil))
            }

        case "notifyCharacteristic":
            if let arguments = call.arguments as? [String: Any],
               let deviceIdString = call.arguments["deviceId"] as? String,
               let serviceUuidString = arguments["serviceUuid"] as? String,
               let characteristicUuid = arguments["characteristicUuid"] as? String,
               let deviceId = UUID(uuidString: deviceIdString) {
                let serviceUuid = CBUUID(string: serviceUuidString)
                manager.listenForNotifications(deviceId: deviceId, serviceUUID: serviceUuid, characteristicUUID: characteristicUuid) { dataResult in
                    switch dataResult {
                    case .success(let healthData):
                        result([
                            "type": healthData.type == .heartRate ? "HEART_RATE" : "GLUCOSE",
                            "value": healthData.value,
                            "timestamp": Int64(healthData.timestamp.timeIntervalSince1970 * 1000)
                        ])
                    case .failure(let error):
                        result(FlutterError(code: "NOTIFY_FAILED", message: error.localizedDescription, details: nil))
                    }
                }
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing or invalid arguments", details: nil))
            }

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/com/vishal/ble_health_device_plugin/BleHealthDevicePlugin.kt
package com.vishal.ble_health_device_plugin

import android.content.Context
import androidx.annotation.NonNull
import com.example.smart_ble_health_device_sdk.DeviceConfiguration
import com.example.smart_ble_health_device_sdk.DeviceType
import com.example.smart_ble_health_device_sdk.HealthBLEManager
import com.example.smart_ble_health_device_sdk.devices.HealthData
import com.example.smart_ble_health_device_sdk.model.HealthDataType
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.*

class BleHealthDevicePlugin : FlutterPlugin, MethodCallHandler {
  private lateinit var channel: MethodChannel
  private lateinit var context: Context
  private lateinit var manager: HealthBLEManager

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    context = flutterPluginBinding.applicationContext
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "ble_health_device_plugin")
    channel.setMethodCallHandler(this)
    val config = listOf(
      DeviceConfiguration(
        deviceId = UUID.randomUUID(),
        deviceType = DeviceType.GLUCOSE,
        allowedOperations = listOf()
      )
    )
    manager = HealthBLEManager(context, config)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "getPlatformVersion" -> {
        result.success("Android ${android.os.Build.VERSION.RELEASE}")
      }
      "startScanning" -> {
        val serviceUuids = (call.argument<List<String>>("serviceUuids") ?: emptyList()).map { UUID.fromString(it) }
        manager.startScanning(serviceUuids) { scanResult ->
          scanResult.onSuccess { devices ->
            val deviceList = devices.map { mapOf("id" to it.id.toString(), "name" to it.name, "isConnected" to it.isConnected) }
            result.success(deviceList)
          }.onFailure { result.error("SCAN_FAILED", it.message, null) }
        }
      }
      "connect" -> {
        val deviceId = call.argument<String>("deviceId")
        if (deviceId != null) {
          manager.connect(deviceId) { connectResult ->
            connectResult.onSuccess { result.success(null) }
              .onFailure { result.error("CONNECT_FAILED", it.message, null) }
          }
        } else {
          result.error("INVALID_ARGUMENT", "Device ID is required", null)
        }
      }
      "disconnect" -> {
        val deviceId = call.argument<String>("deviceId")
        if (deviceId != null) {
          val success = manager.disconnect(deviceId)
          if (success) {
            result.success(null)
          } else {
            result.error("DISCONNECT_FAILED", "Failed to disconnect", null)
          }
        } else {
          result.error("INVALID_ARGUMENT", "Device ID is required", null)
        }
      }
      "readCharacteristic" -> {
        val deviceId = call.argument<String>("deviceId")
        val serviceUuid = call.argument<String>("serviceUuid")?.let { UUID.fromString(it) }
        val characteristicUuid = call.argument<String>("characteristicUuid")
        if (deviceId != null && serviceUuid != null && characteristicUuid != null) {
          manager.readCharacteristic(deviceId, serviceUuid, characteristicUuid) { dataResult ->
            dataResult.onSuccess { healthData ->
              result.success(
                mapOf(
                  "type" to healthData.type.name,
                  "value" to healthData.value,
                  "timestamp" to healthData.timestamp.time
                )
              )
            }.onFailure { result.error("READ_FAILED", it.message, null) }
          }
        } else {
          result.error("INVALID_ARGUMENT", "Missing required arguments", null)
        }
      }
      "writeCharacteristic" -> {
        val deviceId = call.argument<String>("deviceId")
        val serviceUuid = call.argument<String>("serviceUuid")?.let { UUID.fromString(it) }
        val characteristicUuid = call.argument<String>("characteristicUuid")
        val type = call.argument<String>("type")?.let { HealthDataType.valueOf(it) }
        val value = call.argument<Int>("value")
        val timestamp = call.argument<Int>("timestamp")?.toLong()
        if (deviceId != null && serviceUuid != null && characteristicUuid != null && type != null && value != null && timestamp != null) {
          val healthData = HealthData(type, value, Date(timestamp))
          manager.writeCharacteristic(serviceUuid, deviceId, characteristicUuid, healthData) { writeResult ->
            writeResult.onSuccess { result.success(null) }
              .onFailure { result.error("WRITE_FAILED", it.message, null) }
          }
        } else {
          result.error("INVALID_ARGUMENT", "Missing required arguments", null)
        }
      }
      "notifyCharacteristic" -> {
        val deviceId = call.argument<String>("deviceId")
        val serviceUuid = call.argument<String>("serviceUuid")?.let { UUID.fromString(it) }
        val characteristicUuid = call.argument<String>("characteristicUuid")
        if (deviceId != null && serviceUuid != null && characteristicUuid != null) {
          manager.notifyCharacteristic(deviceId, serviceUuid, characteristicUuid) { dataResult ->
            dataResult.onSuccess { healthData ->
              result.success(
                mapOf(
                  "type" to healthData.type.name,
                  "value" to healthData.value,
                  "timestamp" to healthData.timestamp.time
                )
              )
            }.onFailure { result.error("NOTIFY_FAILED", it.message, null) }
          }
        } else {
          result.error("INVALID_ARGUMENT", "Missing required arguments", null)
        }
      }
      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}